source/common/buffer/buffer_impl.cc

```diff
@@ -339,55 +300,113 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
+Reservation OwnedImpl::reserveApproximately(uint64_t length) {
+  Reservation reservation = Reservation::bufferImplUseOnlyConstruct(*this);
+  if (length == 0) {
+    return reservation;
   }
-  // Check whether there are any empty slices with reservable space at the end of the buffer.
-  size_t first_reservable_slice = slices_.size();
-  while (first_reservable_slice > 0) {
-    if (slices_[first_reservable_slice - 1].reservableSize() == 0) {
-      break;
-    }
-    first_reservable_slice--;
-    if (slices_[first_reservable_slice].dataSize() != 0) {
-      // There is some content in this slice, so anything in front of it is non-reservable.
-      break;
-    }
+
+  // Remove any empty slices at the end.
+  while (!slices_.empty() && slices_.back().dataSize() == 0) {
+    slices_.pop_back();
   }
 
-  // Having found the sequence of reservable slices at the back of the buffer, reserve
-  // as much space as possible from each one.
-  uint64_t num_slices_used = 0;
   uint64_t bytes_remaining = length;
-  size_t slice_index = first_reservable_slice;
-  while (slice_index < slices_.size() && bytes_remaining != 0 && num_slices_used < num_iovecs) {
-    auto& slice = slices_[slice_index];
-    const uint64_t reservation_size = std::min(slice.reservableSize(), bytes_remaining);
-    if (num_slices_used + 1 == num_iovecs && reservation_size < bytes_remaining) {
-      // There is only one iovec left, and this next slice does not have enough space to
-      // complete the reservation. Stop iterating, with last one iovec still unpopulated,
-      // so the code following this loop can allocate a new slice to hold the rest of the
-      // reservation.
+  uint64_t reserved = 0;
+  auto& reservation_slices = reservation.bufferImplUseOnlySlices();
+  auto& reservation_owned_slices = reservation.bufferImplUseOnlyOwnedSlices();
+
+  // Check whether there are any empty slices with reservable space at the end of the buffer.
+  uint64_t reservable_size = slices_.empty() ? 0 : slices_.back().reservableSize();
+  if (reservable_size >= length || reservable_size >= (Slice::default_slice_size_ / 8)) {
+    auto& last_slice = slices_.back();
+    const uint64_t reservation_size = std::min(last_slice.reservableSize(), bytes_remaining);
+    auto slice = last_slice.reserve(reservation_size);
+    reservation_slices.push_back(slice);
+    reservation_owned_slices.push_back(nullptr);
+    bytes_remaining -= slice.len_;
+    reserved += slice.len_;
+  }
+
+  while (bytes_remaining != 0) {
+    const uint64_t size = Slice::default_slice_size_;
+
+    // If the next slice would go over the desired size, and the amount already reserved is already
+    // at least one full slice in size, stop allocating slices. This prevents returning a
+    // reservation larger than requested, which could go above the watermark limits for a watermark
+    // buffer, unless the size would be very small (less than 1 full slice).
+    if (size > bytes_remaining && reserved >= size) {
```

Comment:
Removing the preferred_length argument seems like an improvement since it would avoid the situation where the reservation is smaller than the preferred size.  The 128kb default feels large, but it's not much of an issue if buffer limits are explicitly configured.